doc/developer/design/20231117_copy_to_s3.md

```diff
@@ -0,0 +1,303 @@
+the unit as per their documentation below.
+
+> Numeric with Unit: Some numeric parameters have an implicit unit, because they describe
+ quantities of memory or time. The unit might be bytes, kilobytes, blocks (typically eight
+ kilobytes), milliseconds, seconds, or minutes. An unadorned numeric value for one of these
+ settings will use the setting's default unit, which can be learned from pg_settings.unit.
+ For convenience, settings can be given with a unit specified explicitly, for example '120 ms' for
+ a time value, and they will be converted to whatever the parameter's actual unit is. Note that the
+ value must be written as a string (with quotes) to use this feature. The unit name is
+ case-sensitive, and there can be whitespace between the numeric value and the unit.
+
+> Valid memory units are B (bytes), kB (kilobytes), MB (megabytes), GB (gigabytes), and TB
+(terabytes). The multiplier for memory units is 1024, not 1000.
+
+We can take a combination of both the approaches of just accepting the MAX FILE SIZE in bytes
+along with supporting a string like `'5 GB'` similar to Redshift/Postgres.
+
+So the `MAX FILE SIZE` can be either an integer or a parseable string and this will be
+Postgres compatible as well. Note: Even though Postgres specifies MB not MiB, GB not GiB, the
+multiplier is 1024 not 1000. To keep things consistent we should have the same multipliers
+and document it.
+
+### User Experience
+- User will need either an AWS IAM user (to use credentials) or a AWS IAM role (to use AssumeRole)
+on their end to set up an [AWS connection](../design/20231110_aws_connections.md).
+- User needs to create the AWS connection in Materialize to use for `COPY TO ... s3://...`
+ and check with `VALIDATE CONNECTION` that it's working.
+- User will also need to give the AWS user/role for Materialize the following access to the
+s3 prefixes so that we can write the data.
+  - `s3:ListBucket`: We need this permission to make sure that we are writing to an empty path.
+  - `s3:PutObject`: This is required to actually upload files to S3.
+- If they try to run a `COPY` to S3 without all the permissions, then the operation should fail
+with an error. Note, running a `VALIDATE CONNECTION` does not guarantee that the specific S3
+permissions are set up correctly. These can only be validated when the actual `COPY` command
+is run.
+- If the S3 path given already has some data in it, we should fail the operation.
+- When they run the `COPY` command, to keep things simple we'll block the UI till the operation
+is complete, similar to any other query.
+- If successful, the output of the `COPY` command should return the number of rows written to S3
+similar to a [Postgres `COPY` output](https://www.postgresql.org/docs/current/sql-copy.html) of
+`COPY <row_count>`.
+- If the operation fails midway for some reason or is cancelled by the user, we should show
+a notice to the users that incomplete data could have been written to the S3 path and will
```

Comment:
If a user incorrectly handles or loses the result of the `COPY` command, there is a chance they won't know if the path they have in S3 is incomplete or not.

Would it make sense to introduce some sort of sentinel file to indicate a complete copy has occurred? Either something we write at the beginning (e.g. an empty blob named `INCOMPLETE`) that is deleted at the end of the copy, or a `COMPLETE` blob written at the end to indicate full completion? The latter would be easier since it wouldn't require the DeleteObject permission.

This wouldn't be that critical, but might help simplify confusion when errors do occur.

Reply:
Yeah, there's an open question around having a manifest file at the end https://github.com/MaterializeInc/materialize/pull/23556/files#diff-e90b6157a78c239d2480f26f26cd737f66ea6813bd1fc149294b23b9f76d0e68R297 which has the same purpose. IMO it's better to write a file at the end instead of removing, exactly as you mentioned it would not require DeleteObject.